rplugin/python3/deoplete/sources/deoplete_jedi.py

```diff
@@ -47,21 +57,120 @@ def get_complete_position(self, context):
         m = re.search(r'\w*$', context['input'])
         return m.start() if m else -1
 
+    def indent_bounds(self, line, source):
+        """Gets the bounds for a Python block.
+
+        Only search for def or class for context.  Jedi only returns results
+        from the lines before the current one.
+        """
+        start = line - 1
+        indent = len(source[start]) - len(source[start].lstrip())
+
+        for i in range(start, 0, -1):
+            s_line = source[i]
+            if not s_line.strip() or not _block_re.match(s_line):
+                continue
+
+            line_indent = len(s_line) - len(s_line.lstrip())
+            if line_indent < indent:
+                start = i
+                indent = line_indent
+                break
+
+        return [start, line - 1]
+
     def gather_candidates(self, context):
         line = context['position'][1]
         col = context['complete_position']
         buf = self.vim.current.buffer
-        source = '\n'.join(buf[:])
+        source = buf[:]
         cline = self.vim.current.line
 
+        extra_modules = []
+        cache_key = None
+        deoplete_input = context['input'].strip()
+        cache_line = 0
+
+        if self.cache_enabled:
+            # Caching based on context input.  If the input is blank, it was
+            # triggered with `.` to get module completions.
+            #
+            # The module files as reported by Jedi are stored with their
+            # modification times to help detect if a cache needs to be
+            # refreshed.
+            #
+            # For scoped variables in the buffer, construct a cache key using
+            # the filename and a partial input prefix.  The buffer file's
+            # modification time is checked to see if the completion needs to be
+            # refreshed.  The approximate scope lines are also cached to help
+            # invalidate the cache based on line position.
+
+            if context.get('complete_str'):
+                # Note: Module completions will be an empty string.
+                #
+                # Jedi filters the results making making the cache impractical
+                # for buffer variables.  Force Jedi to complete a partial
+                # string based on deoplete.nvim's minimum start length.
+                #
+                # Typing `ose` will return only return matches that start with
+                # `ose.*`.  Typing `os` will return matches that start with
+                # `os.*`, which includes `ose.*`.  Without forcing the minimum
+                # length, we would end up too many cache entries with
+                # duplicates.
+                #
+                # Another reason is deoplete.nvim does fuzzy matching.  Without
+                # this, completing `oseo` would give no results from Jedi.
+                # With the partially cached `os` entry, it will be able to
+                # match `OSError`.
```

Comment:
I just looked closer and it seems like I'm incorrect about Jedi filtering the results.  Taking another look, but it can be a separate PR if you want to close the issue.